decode.go

```diff
@@ -0,0 +1,69 @@
+package fetcher
+
+import (
+	"context"
+	"encoding/gob"
+	"encoding/json"
+	"encoding/xml"
+	"io"
+)
+
+// Decoder is an interface for providing a custom decoder
+type DecodeFunc func(io.Reader, interface{}) error
+
+func jsonDecodeFunc(r io.Reader, v interface{}) error {
+	return json.NewDecoder(r).Decode(v)
+}
+
+func gobDecodeFunc(r io.Reader, v interface{}) error {
+	return gob.NewDecoder(r).Decode(v)
+}
+
+func xmlDecodeFunc(r io.Reader, v interface{}) error {
+	return xml.NewDecoder(r).Decode(v)
+}
+
+// DecodeOption is a func to configure optional Response settings
+type DecodeOption func(c context.Context, resp *Response) error
+
+// DecodeWithJSON json decodes the body of the Response
+func DecodeWithJSON() DecodeOption {
+	return func(c context.Context, resp *Response) error {
+		resp.decodeFunc = jsonDecodeFunc
+		return nil
+	}
+}
+
+// DecodeWithGob gob decodes the body of the Response
+func DecodeWithGob() DecodeOption {
+	return func(c context.Context, resp *Response) error {
+		resp.decodeFunc = gobDecodeFunc
+		return nil
+	}
+}
+
+// DecodeWithXML xml decodes the body of the Response
+func DecodeWithXML() DecodeOption {
+	return func(c context.Context, resp *Response) error {
+		resp.decodeFunc = xmlDecodeFunc
+		return nil
+	}
+}
+
+// DecodeWithCopiedBody makes a copy of the body available in
+func DecodeWithCopiedBody() DecodeOption {
+	return func(c context.Context, resp *Response) error {
+		buf := getBuffer()
```

Comment:
In order for Close() to run putBuffer, you'll need to also set resp.keepBody = true here.